File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Services/ReviewService.cs
using Ecommerce.Dto;
using Ecommerce.Models;
using Ecommerce.Repositories;

namespace Ecommerce.Services;

public class ReviewService
{
    private readonly ReviewRepository _reviewRepository;
    private readonly ProductService _productService;
    private readonly AuthService _authService;

    public ReviewService(
        ReviewRepository reviewRepository,
        ProductService productService,
        AuthService authService)
    {
        _reviewRepository = reviewRepository;
        _productService = productService;
        _authService = authService;
    }

    public async Task<List<ReviewDto>> FindByProductAsync(long productId)
    {
        _ = await _productService.GetProductAsync(productId);
        var reviews = await _reviewRepository.FindByProductIdAsync(productId);
        return reviews.Select(ToDto).ToList();
    }

    public async Task<ReviewDto> CreateAsync(long productId, int rating, string? comment)
    {
        var user = await _authService.CurrentUserAsync();
        var product = await _productService.GetProductAsync(productId);

        if (await _reviewRepository.FindByUserIdAndProductIdAsync(user.Id, productId) is not null)
        {
            throw new ArgumentException("You have already reviewed this product");
        }

        var review = new Review
        {
            UserId = user.Id,
            ProductId = product.Id,
            Rating = rating,
            Comment = comment
        };

        review = await _reviewRepository.SaveAsync(review);
        review.User = user;
        return ToDto(review);
    }

    private static ReviewDto ToDto(Review review) =>
        new(
            review.Id,
            review.UserId,
            review.User.Name,
            review.ProductId,
            review.Rating,
            review.Comment,
            review.CreatedAt);
}

File: rgctl-tests/ecommerce-java/src/main/java/com/example/ecommerce/service/ReviewService.java
package com.example.ecommerce.service;

import com.example.ecommerce.dto.ReviewDto;
import com.example.ecommerce.entity.Product;
import com.example.ecommerce.entity.Review;
import com.example.ecommerce.entity.User;
import com.example.ecommerce.repository.ReviewRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class ReviewService {

    private final ReviewRepository reviewRepository;
    private final ProductService productService;
    private final AuthService authService;

    public ReviewService(ReviewRepository reviewRepository, ProductService productService, AuthService authService) {
        this.reviewRepository = reviewRepository;
        this.productService = productService;
        this.authService = authService;
    }

    @Transactional(readOnly = true)
    public List<ReviewDto> findByProduct(Long productId) {
        productService.getProduct(productId);
        return reviewRepository.findByProductId(productId).stream().map(this::toDto).toList();
    }

    @Transactional
    public ReviewDto create(Long productId, int rating, String comment) {
        User user = authService.currentUser();
        Product product = productService.getProduct(productId);
        if (reviewRepository.findByUserIdAndProductId(user.getId(), productId).isPresent()) {
            throw new IllegalArgumentException("You have already reviewed this product");
        }
        Review review = new Review();
        review.setUser(user);
        review.setProduct(product);
        review.setRating(rating);
        review.setComment(comment);
        return toDto(reviewRepository.save(review));
    }

    private ReviewDto toDto(Review review) {
        return new ReviewDto(
                review.getId(),
                review.getUser().getId(),
                review.getUser().getName(),
                review.getProduct().getId(),
                review.getRating(),
                review.getComment(),
                review.getCreatedAt()
        );
    }
}
